File: Application/Sources/Helpers/StoreReview.m
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

#import "StoreReview.h"

#import "PlaySRG-Swift.h"

@import StoreKit;

@implementation StoreReview

#pragma mark Class methods

+ (void)requestReview
{
#if !defined(DEBUG) && !defined(NIGHTLY) && !defined(BETA)
    static NSString * const kRequestCountUserDefaultsKey = @"PlaySRGStoreReviewRequestCount";
    NSInteger requestCount = [NSUserDefaults.standardUserDefaults integerForKey:kRequestCountUserDefaultsKey] + 1;
    static const NSInteger kRequestCountThreshold = 50;
    if (requestCount >= kRequestCountThreshold) {
        UIWindowScene *windowScene = UIApplication.sharedApplication.mainWindowScene;
        [SKStoreReviewController requestReviewInScene:windowScene];
        requestCount = 0;
    }
    [NSUserDefaults.standardUserDefaults setInteger:requestCount forKey:kRequestCountUserDefaultsKey];
    [NSUserDefaults.standardUserDefaults synchronize];
#endif
}
@end

File: Application/Sources/Helpers/StoreReview.swift
//
//  Copyright (c) SRG SSR. All rights reserved.
//
//  License information is available from the LICENSE file.
//

import Foundation
import StoreKit

@objc class StoreReview: NSObject {
    /**
     *  Gracefully request for an AppStore review.
     *
     *  @see `SKStoreReviewController` documentation for more information.
     */
    @objc static func requestReview() {
        #if !DEBUG && !NIGHTLY && !BETA
            let userDefaultsKey = "PlaySRGStoreReviewRequestCount"
            let userDefaults = UserDefaults.standard
            var requestCount = userDefaults.integer(forKey: userDefaultsKey) + 1
            let requestCountThreshold = 50

            if requestCount >= requestCountThreshold, let mainWindowScene = UIApplication.shared.mainWindowScene {
                SKStoreReviewController.requestReview(in: mainWindowScene)
                requestCount = 0
            }

            userDefaults.set(requestCount, forKey: userDefaultsKey)
            userDefaults.synchronize()
        #endif
    }
}
